TM-4 Task List Application - Create base

TomsBemis opened the pull request:
**TM-5 Creating new task with create task form component:**
<img width="575" alt="TM-5 PR Create" src="https://github.com/TomsBemis/task-manager-frontend/assets/149372128/6bdd0078-ad58-4196-961e-dcb8fdc6e8d5">
<img width="572" alt="TM-5 PR Create after" src="https://github.com/TomsBemis/task-manager-frontend/assets/149372128/71d2f08e-6a38-4f22-91dc-b5c016b9fccd">

**TM-6 Task list component that is populated by a JSON file in the project:**
<img width="704" alt="TM-6 JSON" src="https://github.com/TomsBemis/task-manager-frontend/assets/149372128/28ea4c7c-8b5d-4670-8ad6-966b5b17e53b">
<img width="573" alt="TM-6 after" src="https://github.com/TomsBemis/task-manager-frontend/assets/149372128/c8fe1573-81bc-4571-ac1c-1c21f10ecba6">

**TM-7 Delete functionality for tasks with delete button in each task item:**
<img width="863" alt="TM-7 Delete before" src="https://github.com/TomsBemis/task-manager-frontend/assets/149372128/15333a5e-5bdc-48ac-8cde-ef70b7eaab74">
<img width="893" alt="TM-7 Delete after" src="https://github.com/TomsBemis/task-manager-frontend/assets/149372128/c7bdc6f5-9345-4fac-a1ca-f48ab40be547">
![image](https://github.com/TomsBemis/task-manager-frontend/assets/149372128/0e2677be-b4da-45ea-808d-bd56cce1103e)

TomsBemis committed:
TM-4 Task model and basic components

Components for creating tasks and listing them in the same page. Deleting a task not implemented

TomsBemis committed:
TM-5 Inputs in create form use element references

TomsBemis committed:
TM-6 Task list component that displays list of tasks from imported JSON file using Task Service

TomsBemis committed:
Merge branch 'TM-6-display-list-of-tasks' into TM-5-create-new-task

TomsBemis committed:
TM-5 TaskCreated event refactored to use Task Service

TomsBemis committed:
TM-4 Removed unused imports and commented the code

TomsBemis committed:
TM-4 Missing task attribute - type

TomsBemis committed:
TM-4 Task list model containing an array of task models

wanpii commented on src/app/tasks/create-task/create-task.component.html:
```diff
@@ -0,0 +1,22 @@
+    <div class="col-xs-6">
+        <div class="panel panel-default">
+            <div class="panel-heading">
+                Create new task
```
Please use ReactiveForms from angular :)

wanpii commented on src/app/tasks/create-task/create-task.component.ts:
```diff
@@ -0,0 +1,30 @@
+
+  constructor(private taskService: TaskService) {}
+
+  onCreateTask(taskTitleInput: HTMLInputElement, taskDescriptionInput: HTMLInputElement, taskTypeInput: HTMLInputElement, taskStatusInput: HTMLInputElement) {
```
Same here, will be much easier and code will be reactive

wanpii commented on src/app/tasks/task-list.model.ts:
```diff
@@ -0,0 +1,7 @@
+import { Task } from "./task.model";
+
+export class TaskList {
+    constructor(
```
Use types or interfaces instead of classes

wanpii replied:
still here, instead of defining classes just define a type or an interface for example 
export type Task = {
id: string,
name: string,
createOn: string
etc
}

wanpii commented on src/app/tasks/task-list/task-list.component.ts:
```diff
@@ -0,0 +1,24 @@
+})
+export class TaskListComponent {
+
+  @Input() taskList!: TaskList|undefined;
```
Don’t use non null assertion (!) you can define the input like this: 
taskList: TaskList | null = null

wanpii commented on src/app/tasks/task.model.ts:
```diff
@@ -0,0 +1,9 @@
+export class Task {
+    constructor(
+        public title: string, 
+        public description: string, 
```
Same here type or interface

wanpii replied:
No need to create a class, just the interface here is enough

wanpii commented on src/app/tasks/task.service.ts:
```diff
@@ -0,0 +1,52 @@
+
+export class TaskService {
+
+    taskCreated = new EventEmitter<Task>();
```
You can define event emitters in the components themselves with angular @Output then use that to call the service method to do the actions you need to

wanpii replied:
So you would create the outputs in task item component then in the task list component you would listen to that event emitter and call the service method from task list component

wanpii commented on src/app/tasks/tasks.component.ts:
```diff
@@ -0,0 +1,43 @@
+    );
+    this.taskService.taskDeleted.subscribe(
+      (taskTitle: string) => {
+        this.taskService.deleteTask(taskTitle);
```
Don’t forget to unsubscribe

TomsBemis committed:
TM-4 Pull Request 2 requested changes

 - Create task component changed to user ReactiveFormsModule.
 - Models implement interfaces.
 - Non null assertions removed by changing typescript configuration.
 - Events are no longer subscribed to. Emitters are not located in TaskService

wanpii commented on tsconfig.json:
```diff
@@ -3,7 +3,7 @@
   "compilerOptions": {
     "outDir": "./dist/out-tsc",
-    "strict": true,
+    "strict": false,
```
please set this back to true, if you have errors in the console they probably need to be fixed 😄

TomsBemis committed:
TM-4 Pull Request 2 requested changes

TaskModel changed to interface and compiler strict mode enabled

wanpii commented on src/app/tasks/create-task/create-task.component.html:
```diff
@@ -0,0 +1,36 @@
+                <form [formGroup]="createTaskForm" (ngSubmit)="onSubmit()">
+                    <label for="title">Task title</label><br>
+                    <input type="text" id="title" name="title" formControlName="title"><br>
+                    @if (createTaskForm.get('title')!.errors) {
```
use ? instead of !

wanpii commented on src/app/tasks/create-task/create-task.component.ts:
```diff
@@ -0,0 +1,61 @@
+      Validators.required,
+      this.validateTitleUnique.bind(this)
+    ]), //Custom validator for unique title
+    'description': new FormControl(),
```
no need for quotations here

wanpii commented on src/app/tasks/create-task/create-task.component.ts:
```diff
@@ -0,0 +1,61 @@
+
+  onSubmit () {
+    this.newTask.title = this.createTaskForm.get('title')!.value; 
+    this.newTask.description = this.createTaskForm.get('description')!.value; 
```
you can create a task object inside of this method instead of declaring it outside also don't use non-null assertion (!) here instead you can use ?

wanpii commented on src/app/tasks/create-task/create-task.component.ts:
```diff
@@ -0,0 +1,61 @@
+      .flatMap(
+        (task: { title: string; }) => {return task.title}
+      ).indexOf(control.value) !== -1) {
+        console.log("A");
```
remove console log

wanpii commented on src/app/tasks/task-list.model.ts:
```diff
@@ -0,0 +1,9 @@
+    tasks: Task[]
+}
+
+export let emptyTaskList: TaskList = {
```
I think it should be const, no? Since you will not be changing this object anywhere

wanpii commented on src/app/tasks/task.model.ts:
```diff
@@ -0,0 +1,15 @@
+    status: string
+}
+
+export let emptyTask: Task = {
```
same here as above

TomsBemis committed:
TM-4 Pull Request 2 requested changes

Non-null assertion operators replaced with optional operators. Form group keys are not strings. Empty models changed to constants. Refactored where new task object is instantiated.

TomsBemis commented:
@wanpii Please check if changes are required
